### ISSUE TITLE
Advanced form functionality improvement

### ISSUE THREAD

**Issue by sinukaarel** [sendsmaily/sendsmaily-wordpress-plugin#32]
I think there is an issue with the language field in the current implementation.

When an admin uses the advanced tab to generate a form the language is added also.
Form is serialized and saved to DB and used later when generated by the widget.

The language field that is saved is not dynamic but is the one that admin used( don't actually know because it uses [WPML](https://wpml.org/) modules language code. I don't have access to this as it is a paid one.)

So my idea is that we should only allow to change part of the form or only values that can be used statically. (Maybe there is some kind of form builder we can use. Or just disable part of the text you can't edit so the admin sees that it exists.)

The language field is dynamic and generated based on the language the customer currently uses. Also, I wonder how the nonce field works. This should be also dynamic and generated each time form is loaded. Currently, the `language` field only creates confusion there. (I had feedback from our customer support department and the question was that should this field be filled out by the admin.)

The 	`<input type="hidden" name="action" value="smly" />` seems redundant.

We can't assume they use [WPML](https://wpml.org/) plugin for their localization. Maybe we can use `WPLANG` constant as backup. [Localization handbook](https://developer.wordpress.org/apis/handbook/internationalization/localization/#using-localizations) shows the usage of it and `get_locale()` [function](https://developer.wordpress.org/reference/functions/get_locale/) uses that constant also. They probably use different values or syntax. The question is whether we should normalize those for example `fr_FR` to `france` or maybe show values available on the website to the admin. ( Don't know it is possible). The idea is that the admin knows the values of the fields that he can use to filter by language.

TLDR:
- Advanced form should allow only to edit those values that are static. PHP (server) generated values should be initialized when the form loads.
- The `<input type="hidden" name="action" value="smly" />` seems redundant.
- The nonce field is also dynamic and should be generated when the form loads.
- Provide language even if WPML plugin does not exist

**Comment by TomAbel** [comment]
Is the idea of <input type="hidden" name="lang"> to be able to separate customers of different languages in the Smaily dashboard? I think using get_locale() would suffice then again more customer-friendly options like France as opposed to fr_FR would be better.

I just found out success/error <p>'s can also be only generated when the form loads and currently don't. So some things should be made static. 

I already removed the action input and nonce since 2.0 won't be using Ajax - we're posting forms to subdomain.sendsmaily.net

**Comment by sinukaarel** [comment]
>Is the idea of <input type="hidden" name="lang"> to be able to separate customers of different languages in the Smaily dashboard? 

Yes, our Smaily customers would like to send welcome emails based on the language the guest are using in WordPress site. Then they would be able to create filters for opt-in workflow. Something like this:
![image](https://user-images.githubusercontent.com/22918359/72533526-fa528180-387d-11ea-89a3-233ddd07e676.png)

I wonder if the last one fires if the language field is empty... Should be tested. OK, this workflow cant even be saved due to the empty field. So we should provide a `none` value as an empty or some other undefined option. 

Or I think another option would be to provide just a default with no filters. The question is now what comes first if the customer signs up with the Russian language. Is the standard email is sent or a Russian one that is behind a filter?

![image](https://user-images.githubusercontent.com/22918359/72535872-f294dc00-3881-11ea-8a16-237d9fd8ab96.png)


>I already removed the action input and nonce since 2.0 won't be using Ajax - we're posting forms to subdomain.sendsmaily.net

Good :+1:

**Comment by TomAbel** [comment]
Yeah great, exactly as I imagined. I'd like some ideas on what would be the best design for this though (only to edit those values that are static). I can only come up with readonly input fields followed by large textarea, or some javascript solution that relies on cursor location and key.

**Comment by kaittodesk** [comment]
One option would be to use some sort of form builder 3rd party plugin that provides the necessary functionality for modifying the newsletter opt-in form. But:
- which 3rd party plugin should we use?
- should it replace the basic/advanced form functionality with single approach?
- does it support different i18n plugins or methods of language detection?
- can we use it to build forms that post directly to /api/opt-in/? If not, does it support (and can we enforce) some sort of CAPTCHA solution;

Another option I can see is that we change the newsletter opt-in form to work as a HTML template. Basically we provide variables that user can set in the template. For example `{{lang}}` would always render language code that we (somehow) detect, and it is up to the user where he/she uses it. Although this option requires user to have knowledge of HTML.

> The question is whether we should normalize those for example fr_FR to france or maybe show values available on the website to the admin. ( Don't know it is possible). The idea is that the admin knows the values of the fields that he can use to filter by language.

This type of normalization should be done only in Smaily. Data that is sent to Smaily should be machine readable (for example `fr_FR`). This allows Smaily to have a determined input that they can use to choose the correct output on (whether it is `fr_FR`, `France` or `Prantsusmaa`).

> Or I think another option would be to provide just a default with no filters. The question is now what comes first if the customer signs up with the Russian language. Is the standard email is sent or a Russian one that is behind a filter?

One thing that is crucial in understanding workflows is that every branch that emanates from the trigger is tried for execution. When one branch does not have a language filter (only a message sending action) then it will be executed (i.e. message is sent). So an opt-in with Russian language would trigger both "Russian" and "Without filter" branches.

**Comment by TomAbel** [comment]
> So an opt-in with the Russian language would trigger both "Russian" and "Without filter" branches.

Could we just default it to en_US/American then?

**Comment by kaittodesk** [comment]
Default the opt-in form language to `en_US`?

**Comment by TomAbel** [comment]
Yes. If getting locale fails, then use en_US instead of `''` which it is currently

**Comment by kaittodesk** [comment]
We could, but that would depend heavily on the language of the website. If the website is in Estonian and does not have any 3rd party plugin installed, then it would be wrong to start sending `en_US` as the default language.

We should make our best effort to try and detect the language of the website and/or allow the user to set a fallback value.

**Comment by sinukaarel** [comment]
Current approach in our WooCommerce plugin.
```
// Language code if using WPML.
$lang = '';
if ( defined( 'ICL_LANGUAGE_CODE' ) ) {
	$lang = ICL_LANGUAGE_CODE;
	// Language code if using polylang.
} elseif ( function_exists( 'pll_current_language' ) ) {
	$lang = pll_current_language();
} else {
	$lang = get_locale();
	if ( strlen( $lang ) > 0 ) {
		// Remove any value past underscore if exists.
		$lang = explode( '_', $lang )[0];
	}
}
// Add language code.
$data['language'] = $lang;

**Comment by kaittodesk** [comment]
Yup. This should cover most of the installations.

**Comment by TomAbel** [comment]
Would introducing a form builder require including the form builder in the repository or would it be just set as required for Smaily plugin? Like WooCommerce is needed for the WooCommerce plugin.

**Comment by kaittodesk** [comment]
It would be better if form builder plugin is installed separately from this plugin (as it is with WooCommerce).

**Comment by TomAbel** [comment]
I tried making the basic form with a lot of WP form builders and created a short google doc to report my findings. TL;DR Contact Form 7 or Caldera for an intuitive drag & drop editor.
https://docs.google.com/document/d/1YOeyNsBf7gK6n1dPgSspaJGLy4tIu7SaAExMGV5am9w/edit?usp=sharing
I'm guessing some preprocessing has to be done to un-nest input fields and make it readable for the Smaily API. I'm not up to date with this, can we handle this inside our plugin? Is calling one plugin's hooks in another plugin possible?

**Comment by sinukaarel** [comment]
Closing as this is no longer relevant after version 2.0

**Comment by sinukaarel** [comment]
I will open this issue again due to the language field being currently only checked against the WPML plugin. https://github.com/sendsmaily/sendsmaily-wordpress-plugin/blob/6d347a1807b7a6470d82af55dcfe31f625181cb5/html/form/advanced.php#L10

Should at least implement the minima as in WooCommerce plugin has.

> Current approach in our WooCommerce plugin.
> 
> ```
> // Language code if using WPML.
> $lang = '';
> if ( defined( 'ICL_LANGUAGE_CODE' ) ) {
> 	$lang = ICL_LANGUAGE_CODE;
> 	// Language code if using polylang.
> } elseif ( function_exists( 'pll_current_language' ) ) {
> 	$lang = pll_current_language();
> } else {
> 	$lang = get_locale();
> 	if ( strlen( $lang ) > 0 ) {
> 		// Remove any value past underscore if exists.
> 		$lang = explode( '_', $lang )[0];
> 	}
> }
> // Add language code.
> $data['language'] = $lang;
> ```

**Comment by TomAbel** [comment]
Closed with #54 and implemented in 2.2.0.
